File: problems/1091/solution.py
from typing import List
from collections import deque


class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        if grid[0][0] == 1:
            return -1
        rows, cols = len(grid), len(grid[0])
        q = deque([(0, 0)])
        grid[0][0] = -1
        dirs = [(-1, 0), (-1, 1), (0, 1), (1, 1), (1, 0), (1, -1), (0, -1), (-1, -1)]
        steps = 1
        while q:
            size = len(q)
            for _ in range(size):
                r, c = q.popleft()
                if r == rows - 1 and c == cols - 1:
                    return steps
                for dr, dc in dirs:
                    nr, nc = r + dr, c + dc
                    if 0 <= nr < rows and 0 <= nc < cols:
                        if grid[nr][nc] == 0:
                            grid[nr][nc] = -1
                            q.append((nr, nc))
            steps += 1
        return -1

File: problems/1091/solution.cpp
#include <queue>
#include <vector>
using namespace std;

class Solution {
 private:
  int gridBFS(vector<vector<int>>& grid) {
    int rows = grid.size();
    int cols = grid[0].size();

    queue<pair<int, int>> q;
    int dirs[8][2] = {{-1, 0}, {-1, 1}, {0, 1},  {1, 1},
                      {1, 0},  {1, -1}, {0, -1}, {-1, -1}};

    q.push({0, 0});
    grid[0][0] = -1;
    int steps = 1;

    while (!q.empty()) {
      int size = q.size();
      for (int i = 0; i < size; i++) {
        auto [r, c] = q.front();
        q.pop();

        if (r == rows - 1 && c == cols - 1) return steps;

        for (auto& d : dirs) {
          int nr = r + d[0];
          int nc = c + d[1];

          if (nr >= 0 && nr < rows && nc >= 0 && nc < cols) {
            if (grid[nr][nc] == 0) {
              grid[nr][nc] = -1;
              q.push({nr, nc});
            }
          }
        }
      }
      steps++;
    }
    return -1;
  }

 public:
  int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
    if (grid[0][0] == 1) return -1;
    return gridBFS(grid);
  }
};